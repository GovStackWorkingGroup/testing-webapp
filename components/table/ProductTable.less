@import '../../styles/Variables.less';
@import '../../styles/Extends.less';

.table {
  display: flex;
  flex-direction: column;
}

.table-body {
  width: 70vw;
}

.table-results-count {
  height: 52px;
  display: flex;
  align-items: center;
  gap: 5px;
}

.table-results-count > p {
  font-size: 12px;
  color: @text-secondary;
}

.product-table-row-section:extend(.table-row) {
  max-height: 60vh;
  overflow-y: scroll;
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: 1px;
  background-color: @white;
}

.product-table-header:extend(.table-header-text) {
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 4fr 4fr 4fr 4fr 2fr;
  grid-template-rows: 48px;
  border: 1px solid #dfdfea;
  border-radius: 8px 8px 0px 0px;
  align-items: center;
  position: sticky;
}

.product-table-row {
  width: calc(100% + 15px);
  height: 100%;
  display: grid;
  grid-template-columns: 1fr 4fr 4fr 4fr 4fr 2fr;
  border-bottom: 1px solid #dfdfea;
  align-items: center;
}

.product-table-row:hover {
  border-width: 0px 1px 1px 1px;
  border-style: solid;
  border-color: #dfdfea;
  box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.25);
}

.product-table-row-opened {
  background-color: @blue-d-1;
}

.details-arrow {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  font-size: 24px;
  padding-right: 10px;
}

.table-bb-section {
  display: flex;
  align-items: center;
}
.table-bb-section > div:first-of-type:extend(.flex-row) {
  width: 100%;
  align-items: center;
}

.table-bb-image {
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  height: 48px;
  overflow: hidden;
  gap: 10px;
}

.bb-image {
  padding-top: 15px;
  padding-bottom: 15px;
  filter: invert(30%) sepia(96%) saturate(2946%) hue-rotate(221deg)
    brightness(92%) contrast(94%);
}

.overflow-count {
  display: flex;
  align-items: center;
  height: 48px;
  background-color: @blue-10;
  width: 24px;
  height: 24px;
  border-radius: 50%;
}

.overflow-count > p {
  margin: auto;
  color: @accents-and-btn;
}

.product-last-update {
  color: @text-secondary;
}

.data-loader-message {
  text-align: center;
}

.table-error-handling:extend(.table-row) {
  display: flex;
  width: 100%;
  height: @table-row;
  align-items: center;
  text-align: center;
  justify-content: center;
}

.bb-icon-tooltip {
  background-color: @charts-chips-and-perc;
  text-transform: capitalize;
  font-size: 14px;
  padding: 5px 16px;
}
