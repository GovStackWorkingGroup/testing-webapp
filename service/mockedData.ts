export const mockedProductList = [
  {
    _id: {
      testApp: 'Biodex222',
      testSuite: 'Finance',
      sourceBranch: 'venenatis',
    },
    compatibilities: [
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'mobility_management',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
      {
        id: 'ba37d9df-c353-4f79-b86f-6416d5b9ca09',
        buildingBlock: 'client_case_management',
        timestamp: 1645985463000,
        saveTime: 1675370507000,
        testsPassed: 6,
        testsFailed: 4,
      },
      {
        id: 'ba37d9df-c353-4f79-b86f-6416d5b9ca09',
        buildingBlock: 'client_case_management',
        timestamp: 1645985463000,
        saveTime: 1675370507000,
        testsPassed: 6,
        testsFailed: 4,
      },
      {
        id: '503d1764-ec33-4763-bd15-4168996f41b2',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672130537000,
        saveTime: 1657518851000,
        testsPassed: 14,
        testsFailed: 8,
      },
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'mobility_management',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
      {
        id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
        buildingBlock: 'consent_management',
        timestamp: 1647762312000,
        saveTime: 1673582060000,
        testsPassed: 8,
        testsFailed: 5,
      },
    ],
  },
  {
    _id: { testApp: 'Subin', testSuite: 'Technology', sourceBranch: 'a' },
    compatibilities: [],
  },
  {
    _id: {
      testApp: 'Tampflex',
      testSuite: 'Consumer Services',
      sourceBranch: 'libero',
    },
    compatibilities: [
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'mobility_management',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
      {
        id: 'ba37d9df-c353-4f79-b86f-6416d5b9ca09',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1645985463000,
        saveTime: 1675370507000,
        testsPassed: 6,
        testsFailed: 4,
      },
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
      {
        id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
        buildingBlock: 'consent_management',
        timestamp: 1647762312000,
        saveTime: 1673582060000,
        testsPassed: 8,
        testsFailed: 5,
      },
      // {
      //   id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
      //   buildingBlock: 'consent_management',
      //   timestamp: 1647762312000,
      //   saveTime: 1673582060000,
      //   testsPassed: 8,
      //   testsFailed: 5,
      // },
    ],
  },
  {
    _id: {
      testApp: 'Zamit',
      testSuite: 'Consumer Services',
      sourceBranch: 'duis',
    },
    compatibilities: [
      {
        id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1647762312000,
        saveTime: 1673582060000,
        testsPassed: 8,
        testsFailed: 5,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zoolab',
      testSuite: 'Public Utilities',
      sourceBranch: 'sapien',
    },
    compatibilities: [
      {
        id: '53f5dd11-767c-4944-9055-72d5b1f4447c',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1659961473000,
        saveTime: 1670701134000,
        testsPassed: 10,
        testsFailed: 6,
      },
    ],
  },
  {
    _id: {
      testApp: 'Job',
      testSuite: 'Consumer Non-Durables',
      sourceBranch: 'donec',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: {
      testApp: 'Cardify',
      testSuite: 'Consumer Services',
      sourceBranch: 'venenatis',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: { testApp: 'Treeflex', testSuite: 'n/a', sourceBranch: 'porttitor' },
    compatibilities: [
      {
        id: '503d1764-ec33-4763-bd15-4168996f41b2',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672130537000,
        saveTime: 1657518851000,
        testsPassed: 14,
        testsFailed: 8,
      },
    ],
  },
  {
    _id: { testApp: 'Biodex', testSuite: 'Finance', sourceBranch: 'venenatis' },
    compatibilities: [
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'mobility_management',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
    ],
  },
  {
    _id: { testApp: 'Subin', testSuite: 'Technology', sourceBranch: 'a' },
    compatibilities: [
      {
        id: '43e0fe2f-dbcd-44f5-931d-b0981cf98b10',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672527384000,
        saveTime: 1645552798000,
        testsPassed: 4,
        testsFailed: 3,
      },
    ],
  },
  {
    _id: {
      testApp: 'Tampflex',
      testSuite: 'Consumer Services',
      sourceBranch: 'libero',
    },
    compatibilities: [
      {
        id: 'ba37d9df-c353-4f79-b86f-6416d5b9ca09',
        buildingBlock: 'client_case_management',
        timestamp: 1645985463000,
        saveTime: 1675370507000,
        testsPassed: 6,
        testsFailed: 4,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zamit',
      testSuite: 'Consumer Services',
      sourceBranch: 'duis',
    },
    compatibilities: [
      {
        id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
        buildingBlock: 'consent_management',
        timestamp: 1647762312000,
        saveTime: 1673582060000,
        testsPassed: 8,
        testsFailed: 5,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zoolab',
      testSuite: 'Public Utilities',
      sourceBranch: 'sapien',
    },
    compatibilities: [
      {
        id: '53f5dd11-767c-4944-9055-72d5b1f4447c',
        buildingBlock: 'data_collection',
        timestamp: 1659961473000,
        saveTime: 1670701134000,
        testsPassed: 10,
        testsFailed: 6,
      },
    ],
  },
  {
    _id: {
      testApp: 'Job',
      testSuite: 'Consumer Non-Durables',
      sourceBranch: 'donec',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: {
      testApp: 'Cardify',
      testSuite: 'Consumer Services',
      sourceBranch: 'venenatis',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: { testApp: 'Treeflex', testSuite: 'n/a', sourceBranch: 'porttitor' },
    compatibilities: [
      {
        id: '503d1764-ec33-4763-bd15-4168996f41b2',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672130537000,
        saveTime: 1657518851000,
        testsPassed: 14,
        testsFailed: 8,
      },
    ],
  },
  {
    _id: { testApp: 'Biodex', testSuite: 'Finance', sourceBranch: 'venenatis' },
    compatibilities: [
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'mobility_management',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
    ],
  },
  {
    _id: { testApp: 'Subin', testSuite: 'Technology', sourceBranch: 'a' },
    compatibilities: [
      {
        id: '43e0fe2f-dbcd-44f5-931d-b0981cf98b10',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672527384000,
        saveTime: 1645552798000,
        testsPassed: 4,
        testsFailed: 3,
      },
    ],
  },
  {
    _id: {
      testApp: 'Tampflex',
      testSuite: 'Consumer Services',
      sourceBranch: 'libero',
    },
    compatibilities: [
      {
        id: 'ba37d9df-c353-4f79-b86f-6416d5b9ca09',
        buildingBlock: 'client_case_management',
        timestamp: 1645985463000,
        saveTime: 1675370507000,
        testsPassed: 6,
        testsFailed: 4,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zamit',
      testSuite: 'Consumer Services',
      sourceBranch: 'duis',
    },
    compatibilities: [
      {
        id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
        buildingBlock: 'consent_management',
        timestamp: 1647762312000,
        saveTime: 1673582060000,
        testsPassed: 8,
        testsFailed: 5,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zoolab',
      testSuite: 'Public Utilities',
      sourceBranch: 'sapien',
    },
    compatibilities: [
      {
        id: '53f5dd11-767c-4944-9055-72d5b1f4447c',
        buildingBlock: 'data_collection',
        timestamp: 1659961473000,
        saveTime: 1670701134000,
        testsPassed: 10,
        testsFailed: 6,
      },
    ],
  },
  {
    _id: {
      testApp: 'Job',
      testSuite: 'Consumer Non-Durables',
      sourceBranch: 'donec',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: {
      testApp: 'Cardify',
      testSuite: 'Consumer Services',
      sourceBranch: 'venenatis',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: { testApp: 'Treeflex', testSuite: 'n/a', sourceBranch: 'porttitor' },
    compatibilities: [
      {
        id: '503d1764-ec33-4763-bd15-4168996f41b2',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672130537000,
        saveTime: 1657518851000,
        testsPassed: 14,
        testsFailed: 8,
      },
    ],
  },
  {
    _id: { testApp: 'Biodex', testSuite: 'Finance', sourceBranch: 'venenatis' },
    compatibilities: [
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'mobility_management',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
    ],
  },
  {
    _id: { testApp: 'Subin', testSuite: 'Technology', sourceBranch: 'a' },
    compatibilities: [
      {
        id: '43e0fe2f-dbcd-44f5-931d-b0981cf98b10',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672527384000,
        saveTime: 1645552798000,
        testsPassed: 4,
        testsFailed: 3,
      },
    ],
  },
  {
    _id: {
      testApp: 'Tampflex',
      testSuite: 'Consumer Services',
      sourceBranch: 'libero',
    },
    compatibilities: [
      {
        id: 'ba37d9df-c353-4f79-b86f-6416d5b9ca09',
        buildingBlock: 'client_case_management',
        timestamp: 1645985463000,
        saveTime: 1675370507000,
        testsPassed: 6,
        testsFailed: 4,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zamit',
      testSuite: 'Consumer Services',
      sourceBranch: 'duis',
    },
    compatibilities: [
      {
        id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
        buildingBlock: 'consent_management',
        timestamp: 1647762312000,
        saveTime: 1673582060000,
        testsPassed: 8,
        testsFailed: 5,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zoolab',
      testSuite: 'Public Utilities',
      sourceBranch: 'sapien',
    },
    compatibilities: [
      {
        id: '53f5dd11-767c-4944-9055-72d5b1f4447c',
        buildingBlock: 'data_collection',
        timestamp: 1659961473000,
        saveTime: 1670701134000,
        testsPassed: 10,
        testsFailed: 6,
      },
    ],
  },
  {
    _id: {
      testApp: 'Job',
      testSuite: 'Consumer Non-Durables',
      sourceBranch: 'donec',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: {
      testApp: 'Cardify',
      testSuite: 'Consumer Services',
      sourceBranch: 'venenatis',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: { testApp: 'Treeflex', testSuite: 'n/a', sourceBranch: 'porttitor' },
    compatibilities: [
      {
        id: '503d1764-ec33-4763-bd15-4168996f41b2',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672130537000,
        saveTime: 1657518851000,
        testsPassed: 14,
        testsFailed: 8,
      },
    ],
  },
  {
    _id: { testApp: 'Biodex', testSuite: 'Finance', sourceBranch: 'venenatis' },
    compatibilities: [
      {
        id: '07112c0a-8263-4717-92ce-c52bca785624',
        buildingBlock: 'mobility_management',
        timestamp: 1644849545000,
        saveTime: 1659380963000,
        testsPassed: 2,
        testsFailed: 2,
      },
    ],
  },
  {
    _id: { testApp: 'Subin', testSuite: 'Technology', sourceBranch: 'a' },
    compatibilities: [
      {
        id: '43e0fe2f-dbcd-44f5-931d-b0981cf98b10',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672527384000,
        saveTime: 1645552798000,
        testsPassed: 4,
        testsFailed: 3,
      },
    ],
  },
  {
    _id: {
      testApp: 'Tampflex',
      testSuite: 'Consumer Services',
      sourceBranch: 'libero',
    },
    compatibilities: [
      {
        id: 'ba37d9df-c353-4f79-b86f-6416d5b9ca09',
        buildingBlock: 'client_case_management',
        timestamp: 1645985463000,
        saveTime: 1675370507000,
        testsPassed: 6,
        testsFailed: 4,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zamit',
      testSuite: 'Consumer Services',
      sourceBranch: 'duis',
    },
    compatibilities: [
      {
        id: '0ba8dcdc-0c12-4a7e-b9c6-e7674379cf15',
        buildingBlock: 'consent_management',
        timestamp: 1647762312000,
        saveTime: 1673582060000,
        testsPassed: 8,
        testsFailed: 5,
      },
    ],
  },
  {
    _id: {
      testApp: 'Zoolab',
      testSuite: 'Public Utilities',
      sourceBranch: 'sapien',
    },
    compatibilities: [
      {
        id: '53f5dd11-767c-4944-9055-72d5b1f4447c',
        buildingBlock: 'data_collection',
        timestamp: 1659961473000,
        saveTime: 1670701134000,
        testsPassed: 10,
        testsFailed: 6,
      },
    ],
  },
  {
    _id: {
      testApp: 'Job',
      testSuite: 'Consumer Non-Durables',
      sourceBranch: 'donec',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: {
      testApp: 'Cardify',
      testSuite: 'Consumer Services',
      sourceBranch: 'venenatis',
    },
    compatibilities: [
      {
        id: '6b696f8f-8d3f-4322-a269-8d49970849f0',
        buildingBlock: 'analytics_and_business_intel',
        timestamp: 1663755505000,
        saveTime: 1670490781000,
        testsPassed: 12,
        testsFailed: 7,
      },
    ],
  },
  {
    _id: { testApp: 'Treeflex', testSuite: 'n/a', sourceBranch: 'porttitor' },
    compatibilities: [
      {
        id: '503d1764-ec33-4763-bd15-4168996f41b2',
        buildingBlock: 'artificial_intelligence',
        timestamp: 1672130537000,
        saveTime: 1657518851000,
        testsPassed: 14,
        testsFailed: 8,
      },
    ],
  },
];
