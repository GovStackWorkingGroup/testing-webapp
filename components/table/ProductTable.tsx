import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import InfiniteScroll from 'react-infinite-scroll-component';
import { getSoftwaresData, getSoftwareListCount } from '../../service/serviceAPI';
import { ProductsType } from '../../service/types';
import InfiniteScrollCustomLoader from '../InfiniteScrollLoader';
import ProductTableHeader from './ProductTableHeader';
import ProductTableRow from './ProductTableRow';
import TableErrorHandling from './TableErrorHandling';
import { SoftwaresTableSortByType } from './types';

const ProductTable = () => {
  const [productsList, setProductsList] = useState<ProductsType[]>([]);
  const [productListLength, setProductListLength] = useState<number>(0);
  const [isErrorFetchingData, setIsErrorFetchingData] = useState<boolean>(false);
  const [isErrorFetchingCount, setIsErrorFetchingCount] = useState<boolean>(false);
  const [sortBy, setSortBy] = useState<SoftwaresTableSortByType>({
    testApp: { field: 'testApp', order: null },
    lastUpdate: { field: 'lastUpdate', order: null },
    overallCompatibility: { field: 'overallCompatibility', order: null }
  });

  console.log(sortBy);

  const { formatMessage } = useIntl();
  const format = useCallback(
    (id: string) => formatMessage({ id }),
    [formatMessage]
  );

  useEffect(() => {
    fetchData();
  }, [sortBy]);

  const fetchData = async () => {
    const [data, count] = await Promise.all([
      getSoftwaresData(0, sortBy),
      getSoftwareListCount(),
    ]);

    if (data.status) {
      setProductsList(data.data);
    } else {
      setIsErrorFetchingData(true);
    }

    if (count.status) {
      setProductListLength(count.data);
    } else {
      setIsErrorFetchingCount(true);
    }
  };

  const handleLoadMoreData = useCallback(async () => {
    const data = await getSoftwaresData(productsList.length, sortBy);

    if (data.status) {
      setProductsList([...productsList, ...data.data]);
    } else {
      setProductsList([...productsList]);
    }
  }, [productsList]);

  const handleSorting = (tableSortProperties: SoftwaresTableSortByType) => {
    setSortBy(tableSortProperties);
  };

  return (
    <div className='table'>
      <div className='table-results-count'>
        <p>
          {isErrorFetchingCount ? format('table.no_result_count.message') : productListLength}
        </p>
        <p>
          {isErrorFetchingCount ? null : productListLength === 1
            ? format('table.result.label')
            : format('table.result.plural.label')}
        </p>
      </div>
      <div className='table-body'>
        <ProductTableHeader handleSorting={handleSorting}/>
        {isErrorFetchingData ? (
          <TableErrorHandling />
        ) : (
          <div id='scrollableDiv' className='product-table-row-section'>
            <InfiniteScroll
              scrollableTarget='scrollableDiv'
              dataLength={productsList?.length ?? 0}
              next={handleLoadMoreData}
              hasMore={
                isErrorFetchingData
                  ? false
                  : productListLength > productsList?.length
              }
              loader={<InfiniteScrollCustomLoader />}
              style={{ overflowX: 'hidden' }}
            >
              {productsList.map((product, key) => (
                <ProductTableRow product={product} key={`product-${key}`} />
              ))}
            </InfiniteScroll>
          </div>
        )}
      </div>
    </div>
  );
};

export default ProductTable;
