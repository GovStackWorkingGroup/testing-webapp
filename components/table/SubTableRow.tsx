import React from 'react';
import { CompatibilitiesType } from '../../service/types';
import BBImage from './BuildingBlocksImage';

type Props = {
  compatibilities: CompatibilitiesType[];
};

const SubTableRow = ({ compatibilities }: Props) => {
  return (
    <div>
      {compatibilities.map((bb, bbIdx) => {
        return (
          <div className='sub-table-row' key={`sub-table-row-${bbIdx}`}>
            <div></div>
            <div className='sub-table-row-bb' data-testid='bb-name'>
              <BBImage imagePath={bb.buildingBlock} />
              <p>{bb.buildingBlock}</p>
            </div>
            <div className='sub-table-content-tests'>
              <div data-testid='tests-passed'>
                <p>{bb.testsPassed}</p>
              </div>
              <div data-testid='tests-failed'>
                <p>{bb.testsFailed}</p>
              </div>
            </div>
            <div className='sub-table-content-compatibility' data-testid='compatibility'>
              <p>{`${(bb.compatibility*100).toFixed(2)}%`}</p>
            </div>
            <div className='test-details-arrow'></div>
          </div>
        );
      })}
    </div>
  );
};

export default SubTableRow;
