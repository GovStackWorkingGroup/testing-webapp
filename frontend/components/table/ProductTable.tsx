import isEqual from 'lodash/isEqual';
import { useCallback, useEffect, useState } from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import { useIntl } from 'react-intl';
import { getSoftwareListCount, getSoftwaresData } from '../../service/serviceAPI';
import { ProductsType } from '../../service/types';
import InfiniteScrollCustomLoader from '../InfiniteScrollLoader';
import ProductTableHeader from './ProductTableHeader';
import ProductTableRow from './ProductTableRow';
import TableErrorHandling from './TableErrorHandling';
import { SoftwaresTableSortByType } from './types';

const defaultSortBy = {
  testApp: { field: 'testApp', order: null },
  lastUpdate: { field: 'lastUpdate', order: null },
  overallCompatibility: { field: 'overallCompatibility', order: null }
} as SoftwaresTableSortByType;

const ProductTable = ({ showAll }: { showAll: boolean }) => {
  const [productsList, setProductsList] = useState<ProductsType[]>([]);
  const [productListLength, setProductListLength] = useState<number>(0);
  const [error, setError] = useState<string | null>(null);
  const [sortBy, setSortBy] = useState<SoftwaresTableSortByType>(defaultSortBy);
  const [loading, setLoading] = useState(true);

  const { formatMessage } = useIntl();
  const format = useCallback((id: string) => formatMessage({ id }), [formatMessage]);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const [data, count] = await Promise.all([
          getSoftwaresData(0, sortBy, 'main', showAll),
          getSoftwareListCount()
        ]);

        if (data.status) {
          setProductsList(data.data);
        }

        if (count.status) {
          setProductListLength(count.data);
        }
      } catch {
        setError('dataFetchFail');
      } finally {
        setLoading(false); // set loading to false once data is fetched or error occurs
      }
    };

    fetchData();
  }, [sortBy]);

  const handleLoadMoreData = useCallback(async () => {
    try {
      const data = await getSoftwaresData(productsList.length, sortBy);

      if (data.status) {
        setProductsList((prev) => [...prev, ...data.data]);
      }
    } catch {
      setError('dataLoadFail');
    }
  }, [productsList, sortBy]);

  const handleSorting = (tableSortProperties: SoftwaresTableSortByType) => {
    if (!isEqual(tableSortProperties, sortBy)) {
      setSortBy(tableSortProperties);
    }
  };

  if (error) {
    return <TableErrorHandling />;
  }

  return (
    <div className='table'>
      <div className='table-results-count'>
        <p>{productListLength}</p>
        <p>{productListLength === 1 ? format('table.result.label') : format('table.result.plural.label')}</p>
      </div>
      <div className='table-body'>
        <ProductTableHeader handleSorting={handleSorting}/>
        <div id='scrollableDiv' className='product-table-row-section'>
          {loading ? (
            <div className="product-table-row-section">
              <InfiniteScrollCustomLoader />
            </div>
          ) : (
            <>
              <InfiniteScroll
                scrollableTarget='scrollableDiv'
                dataLength={productsList.length}
                next={handleLoadMoreData}
                hasMore={productListLength > productsList.length}
                loader={<InfiniteScrollCustomLoader />}
                style={{ overflowX: 'hidden' }}
              >
                {productsList.map((product, index) => (
                  <ProductTableRow product={product} key={index} />
                ))}
              </InfiniteScroll>
            </>
          )}
        </div>
      </div>
    </div>
  );
};

export default ProductTable;
