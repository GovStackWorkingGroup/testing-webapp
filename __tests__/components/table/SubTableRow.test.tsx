import React from 'react';
import SubTableRow from '../../../components/table/SubTableRow';
import { render } from '../../test-utils/test-utils';
import { buildingBlock, mockedProduct } from './mockedData/ProductTable';

describe('Unit tests for SubTableRow component:', () => {
  const BB_NAME_TEST_ID = 'bb-name';
  const TESTS_PASSED_TEST_ID = 'tests-passed';
  const TESTS_FAILED_TEST_ID = 'tests-failed';
  const COMPATIBILITY_TEST_ID = 'compatibility';

  it('should render the SubTableRow component and match snapshot', () => {
    const { container } = render(
      <SubTableRow
        productName={mockedProduct._id.testApp}
        buildingBlock={buildingBlock}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('when have corresponding values', () => {
    const { getByTestId } = render(
      <SubTableRow
        productName={mockedProduct._id.testApp}
        buildingBlock={buildingBlock}
      />
    );

    expect(getByTestId(BB_NAME_TEST_ID)).toHaveTextContent('Payments');
    expect(getByTestId(TESTS_PASSED_TEST_ID)).toHaveTextContent('2');
    expect(getByTestId(TESTS_FAILED_TEST_ID)).toHaveTextContent('2');
    expect(getByTestId(COMPATIBILITY_TEST_ID)).toHaveTextContent('34%');
  });
});
