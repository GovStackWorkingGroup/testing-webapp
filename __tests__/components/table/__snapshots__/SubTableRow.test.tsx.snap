// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for SubTableRow component: should render the SubTableRow component and match snapshot 1`] = `
<div>
  <div>
    <div
      class="sub-table-row"
    >
      <div />
      <div
        class="sub-table-row-bb"
        data-testid="bb-name"
      >
        <img
          alt="Logo for: Building block label"
          class="bb-image"
          data-nimg="1"
          decoding="async"
          height="18"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2FbuildingBlocks%2FBuilding%20block%20label.png&w=48&q=75"
          srcset="/_next/image?url=%2Fimages%2FbuildingBlocks%2FBuilding%20block%20label.png&w=32&q=75 1x, /_next/image?url=%2Fimages%2FbuildingBlocks%2FBuilding%20block%20label.png&w=48&q=75 2x"
          style="color: transparent;"
          width="18"
        />
        <p>
          Building block label
        </p>
      </div>
      <div
        class="sub-table-content-tests"
      >
        <div
          data-testid="tests-passed"
        >
          <p>
            30
          </p>
        </div>
        <div
          data-testid="tests-failed"
        >
          <p>
            2
          </p>
        </div>
      </div>
      <div
        class="sub-table-content-compatibility"
        data-testid="compatibility"
      >
        <p>
          34.55%
        </p>
      </div>
      <div
        class="test-details-arrow"
      />
    </div>
  </div>
</div>
`;
